### PR TITLE
Support for Node 12, some more simpifications

### PR DESCRIPTION
**Pull request by petermuessig**

### CODE REVIEW

**Comment by petermuessig** [comment]
Hi @IObert , don't be afraid - I just reworked and simplified a few code places. Sorry, I love to simplify code and avoid duplicates. But I want to do some more cleanup when I find some time. I verified most of the code already. Should work fine... :-) We need more tests... :-P

File: generators/app/index.js [hunk 1]
```diff
@@ -4,48 +4,45 @@ const chalk = require("chalk");
...
-const fs = require("fs");
 const path = require("path");
+const fs = require("fs");
+const { rmdir } = require("fs").promises;
 
 const { Octokit } = require("@octokit/rest");
 const AdmZip = require("adm-zip");
 
 const generatorOptions = {
-  generator: {
```
Comment (petermuessig) [inline generators/app/index.js:14]: Not needed as it should be only passed as argument

File: generators/app/index.js [hunk 2]
```diff
@@ -71,31 +75,67 @@ module.exports = class extends Generator {
...
   async prompting() {
     this.log(yosay(`Welcome to the ${chalk.red("easy-ui5")} generator!`));
 
+    // create the octokit client to retrieve the generators from GH org
     const octokit = new Octokit({
       userAgent: `${this.rootGeneratorName()}:${this.rootGeneratorVersion()}`,
       auth: this.options.ghAuthToken,
     });
 
-    const reqRepos = await octokit.repos.listForOrg({
```
Comment (petermuessig) [inline generators/app/index.js:82]: Had some issues when quota exceeded for amount of request while running local tests without authentication - so I added proper error handling

File: generators/app/index.js [hunk 3]
```diff
@@ -71,31 +75,67 @@ module.exports = class extends Generator {
...
+      clearInterval(this.busyIndicatorTimer);
+      delete this.busyIndicatorTimer;
+      process.stdout.write(`\n`);
+    }
+  }
+
   async prompting() {
     this.log(yosay(`Welcome to the ${chalk.red("easy-ui5")} generator!`));
 
+    // create the octokit client to retrieve the generators from GH org
```
Comment (petermuessig) [inline generators/app/index.js:102]: sorry for adding so many comments - but this improves the readability of the code a lot!

File: generators/app/index.js [hunk 4]
```diff
@@ -123,17 +163,27 @@ module.exports = class extends Generator {
         generator = generatorRepos[generatorIdx];
       }
 
-      const reqBranch = await octokit.repos.getBranch({
-        owner: this.options.ghOrg,
-        repo: generator.name,
-        branch: generator.default_branch,
-      });
+      // fetch the available branches to retrieve the latest commit SHA
```
Comment (petermuessig) [inline generators/app/index.js:166]: same here - error logging when no branch exists (e.g. for the middleware and the task repos)

File: generators/app/index.js [hunk 5]
```diff
@@ -146,16 +196,22 @@ module.exports = class extends Generator {
...
         // check if the SHA marker exists to know whether the generator is up-to-date or not
         if (!fs.existsSync(shaMarker)) {
-          this.log(`generator in ${generatorPath} is outdated...`);
+          if (this.options.verbose) {
+            this.log(`Generator "${generator.name}" in "${generatorPath}" is outdated...`);
+          }
           // remove if the SHA marker doesn't exist => outdated!
-          fs.rmdirSync(generatorPath, { recursive: true });
+          this._showBusy(`  - Deleting "${generator.name}"`);
+          await rmdir(generatorPath, { recursive: true });
```
Comment (petermuessig) [inline generators/app/index.js:204]: switched to async to see a proper busy indication

File: generators/app/index.js [hunk 6]
```diff
@@ -178,70 +234,113 @@ module.exports = class extends Generator {
...
+    // helper to derive the subcommand
     function deriveSubcommand(namespace) {
       const match = namespace.match(/[^:]+:(.+)/);
       return match ? match[1] : namespace;
     }
 
-    let defaultSubGenerator;
+    // filter the hidden subgenerators already
+    //   -> subgenerators must be found in env as they are returned by lookup!
+    let subGenerators = env
```
Comment (petermuessig) [inline generators/app/index.js:277]: the hidden filtering can be done only once

File: generators/app/index.js [hunk 7]
```diff
@@ -178,70 +234,113 @@ module.exports = class extends Generator {
...
+      this.log(subGenerators
+        .map(sub => {
+          maxLength = Math.max(sub.namespace.length, maxLength);
+          return sub;
         })
-        .map((sub) => {
-          let line = `${deriveSubcommand(sub.namespace)}\t\t`;
-          if (deriveSubcommand(sub.namespace).length < 8) {
-            line += "\t";
+        .reduce((output, sub) => {
```
Comment (petermuessig) [inline generators/app/index.js:292]: reduce is a very effective command to convert an array into a e.g. string (finally a single value)

File: generators/app/index.js [hunk 8]
```diff
@@ -178,70 +234,113 @@ module.exports = class extends Generator {
...
+          const subGenerator = env.get(sub.namespace);
+          const displayName = subGenerator.displayName || "";
+          let line = `  ${deriveSubcommand(sub.namespace).padEnd(maxLength + 2)}`;
+          if (displayName) {
+            line += ` # ${subGenerator.displayName}`;
```
Comment (petermuessig) [inline generators/app/index.js:297]: using hashes to be in sync with Yeoman logging

File: generators/app/index.js [hunk 9]
```diff
@@ -178,70 +234,113 @@ module.exports = class extends Generator {
...
-          );
-        }
-        return !env.get(sub.namespace)?.hidden;
+    // if a subcommand is provided as argument, identify the matching subgenerator
+    // and remove the rest of the subgenerators from the list for later steps
+    if (this.options.subcommand) {
+      const selectedSubGenerator = subGenerators
+        .filter((sub) => {
+          // identify the subgenerator by subcommand
+          return new RegExp(`:${this.options.subcommand}$`).test(sub.namespace);
```
Comment (petermuessig) [inline generators/app/index.js:310]: endswith is more precise than includes

File: generators/app/index.js [hunk 10]
```diff
@@ -178,70 +234,113 @@ module.exports = class extends Generator {
...
+      }
+    }
+
+    // transform the list of the subgenerators and identify the 
+    // default subgenerator for the default selection
+    let defaultSubGenerator;
+    let maxLength = 0;
+    subGenerators = subGenerators
+      .map(sub => {
+        const generator = env.get(sub.namespace);
```
Comment (petermuessig) [inline generators/app/index.js:331]: prepare the required props and calculate the maxLength of the row to finally use padEnd to ensure a harmonized alignment

**Comment by petermuessig** [comment]
Now with nice busy indication when downloading so that the developer gets proper feedback.

**Comment by IObert** [comment]
Thanks a lot for this PR and the great improvements!

**Pull request merged**